docs/design/single-file-unsupported.md

```diff
@@ -0,0 +1,147 @@
+
+## User Experience
+
+The new attribute will allow developers to easily annotate their APIs known to be problematic when used in the context of a single-file application. This new attribute will first be used by the runtime team to annotate low level [APIs known to be incompatible with single-file](https://docs.microsoft.com/en-us/dotnet/core/deploying/single-file#api-incompatibility). Library authors will then rely on analyzers to know which parts of their code would either need to be fixed or marked with `SingleFileUnsupportedAttribute`.
+
+### Example: API writer
+
+Ben is writing a library which he reckons some of its consumers might want to use in the context of a single-file app, so he decides to see if any warning is produced when `PublishSingleFile` is set in the .csproj file. After doing this, the following code gets a diagnostic produced by the single-file analyzer: 
+
+> FileTable.cs(3,12): Single-file warning ILXXXX: 'System.Reflection.Assembly.GetFiles()' will throw for assemblies embedded in a single-file app.
+
+```C#
+public int FileTableLength(Assembly assembly)
+{
+    return assembly.GetFiles().Length;
+}
+```
+
+He then decides to annotate the method so that users who intent to deploy their app as a single-file binary know that `FileTableLength` will not be safe to use, and can either wrap the  method call in a try-catch block or work around using it.
+
+```C#
+[SingleFileUnsupported("This method will throw for single-file apps.", "https://helpurl")]
+public int FileTableLength(Assembly assembly)
+```
+
+### Example: API consumer
+
+Maria is writing an application which keeps track of all times it has ran in a log next to the running assembly. For this purpose she uses `Assembly.Location`, which is marked with the new `SingleFileUnsupportedAttribute`.
+
+```C#
+void RecordApplicationRun()
+{
+    var path = Path.Combine(Path.GetDirectoryName(Assembly.GetExecutingAssembly().CodeBase), "runs.log");
+    using (StreamWriter sw = File.AppendText(path))
+        sw.WriteLine(DateTime.Now.ToString());
+}
+```
+
+After some time she decides to make her application a single-file app, so that she can easily share it with friends. She then adds the `PublishSingleFile` property to the .csproj.
+
+```XML
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+    <TargetFramework>net5.0</TargetFramework>
+    <PublishSingleFile>true</PublishSingleFile>
+  </PropertyGroup>
+
+</Project>
+```
+
+Now that she has expressed the intent to deploy the app as single-file binary, the following diagnostic is shown in her code:
+
+> Logger.cs(x,y): Single-file warning ILXXXX: 'System.Reflection.Assembly.Location' always returns an empty string for assemblies embedded in a single-file app. If the path to the app directory is needed, consider calling 'System.AppContext.BaseDirectory'.
```

Comment:
The sample above uses `Assembly.CodeBase`, but the sample warning talks about `Assembly.Location`.